dotnet/src/SemanticKernel.Core/Memory/Mem0/Mem0Behavior.cs

```diff
@@ -0,0 +1,178 @@
+/// </para>
+/// </remarks>
+[Experimental("SKEXP0130")]
+public sealed class Mem0Behavior : AIContextBehavior
+{
+    private const string DefaultContextPrompt = "## Memories\nConsider the following memories when answering user questions:";
+
+    private readonly string? _applicationId;
+    private readonly string? _agentId;
+    private readonly string? _threadId;
+    private string? _perOperationThreadId;
+    private readonly string? _userId;
+    private readonly bool _scopeToPerOperationThreadId;
+    private readonly string _contextPrompt;
+
+    private readonly Mem0Client _mem0Client;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="Mem0Behavior"/> class.
+    /// </summary>
+    /// <param name="httpClient">The HTTP client used for making requests.</param>
+    /// <param name="options">Options for configuring the component.</param>
+    /// <remarks>
+    /// The base address of the required mem0 service, and any authentication headers, should be set on the <paramref name="httpClient"/>
+    /// already, when passed as a parameter here. E.g.:
+    /// <code>
+    /// using var httpClient = new HttpClient();
+    /// httpClient.BaseAddress = new Uri("https://api.mem0.ai");
+    /// httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Token", "&lt;Your APIKey&gt;");
+    /// new Mem0Client(httpClient);
+    /// </code>
+    /// </remarks>
+    public Mem0Behavior(HttpClient httpClient, Mem0BehaviorOptions? options = default)
+    {
+        Verify.NotNull(httpClient);
+
+        if (string.IsNullOrWhiteSpace(httpClient.BaseAddress?.AbsolutePath))
+        {
+            throw new ArgumentException("The BaseAddress of the provided httpClient parameter must be set.", nameof(httpClient));
+        }
+
+        this._applicationId = options?.ApplicationId;
+        this._agentId = options?.AgentId;
+        this._threadId = options?.ThreadId;
+        this._userId = options?.UserId;
+        this._scopeToPerOperationThreadId = options?.ScopeToPerOperationThreadId ?? false;
+        this._contextPrompt = options?.ContextPrompt ?? DefaultContextPrompt;
+
+        this._mem0Client = new(httpClient);
+    }
+
+    /// <inheritdoc/>
+    public override Task OnThreadCreatedAsync(string? threadId, CancellationToken cancellationToken = default)
+    {
+        this.ValidatePerOperationThreadId(threadId);
+
+        this._perOperationThreadId ??= threadId;
+        return Task.CompletedTask;
+    }
+
+    /// <inheritdoc/>
+    public override async Task OnNewMessageAsync(string? threadId, ChatMessage newMessage, CancellationToken cancellationToken = default)
+    {
+        Verify.NotNull(newMessage);
+        this.ValidatePerOperationThreadId(threadId);
+
+        switch (newMessage.Role)
+        {
+            case ChatRole u when u == ChatRole.User:
+            case ChatRole a when a == ChatRole.Assistant:
+            case ChatRole s when s == ChatRole.System:
```

Comment:
I guess you can do this...

```suggestion
            case ChatRole when ChatRole.User or ChatRole.Assistant or ChatRole.System:
```

Reply:
Unfortunately, that doesn't work.  ChatRole.User, Assitant and System are properties that are instances of ChatRole with a type, so you have to do checks like this.